src/components/Editor/DrawingBoard/Canvas/Container.ts

```diff
@@ -0,0 +1,216 @@
+
+    if (this.worker.type === tool || tool === ToolType.None) {
+      return;
+    }
+
+    this.worker.flushTask();
+
+    if (tool === ToolType.Line) {
+      this.worker = new LineWorker(this.update, this.emit);
+    } else if (tool === ToolType.Eraser) {
+      this.worker = new EraserWorker(this.update, this.emit);
+    } else if (tool === ToolType.Rect) {
+      this.worker = new RectWorker(this.update, this.emit);
+    } else if (tool === ToolType.Selector) {
+      this.worker = new SelectorWorker(this.update, this.emit);
+    } else {
+      throw new Error(`Undefined tool: ${tool}`);
+    }
+  }
+
+  setMouseClass(tool: ToolType) {
+    this.upperWrapper.getCanvas().className = 'canvas canvas-upper';
+
+    if (tool === ToolType.Line || tool === ToolType.Rect) {
+      this.upperWrapper.getCanvas().classList.add('crosshair', 'canvas-touch-none');
+    } else if (tool === ToolType.Eraser) {
+      this.upperWrapper.getCanvas().classList.add('eraser', 'canvas-touch-none');
+    } else if (tool === ToolType.Selector) {
+      this.upperWrapper.getCanvas().classList.add('canvas-touch-none');
+    }
+  }
+
+  getPointFromTouchyEvent(evt: TouchyEvent): Point {
+    let originY;
+    let originX;
+    if (window.TouchEvent && evt instanceof TouchEvent) {
+      originY = evt.touches[0].clientY;
+      originX = evt.touches[0].clientX;
+    } else {
+      originY = evt.clientY;
+      originX = evt.clientX;
+    }
+    originY += window.scrollY;
+    originX += window.scrollX;
+    return {
+      y: originY - this.offsetY,
+      x: originX - this.offsetX,
+    };
+  }
+
+  onMouseDown(evt: TouchyEvent) {
+    touchy(this.upperWrapper.getCanvas(), addEvent, 'mousemove', this.onMouseMove);
+    this.dragStatus = DragStatus.Drag;
+
+    const point = this.getPointFromTouchyEvent(evt);
+
+    this.worker.mousedown(point, { color: this.color });
+  }
+
+  onMouseMove(evt: TouchyEvent) {
+    const point = this.getPointFromTouchyEvent(evt);
+    if (this.isOutside(point)) {
+      this.onMouseUp();
+      return;
+    }
+
+    if (this.dragStatus === DragStatus.Stop) {
+      return;
+    }
+
+    this.worker.mousemove(point);
+  }
+
+  onMouseUp() {
+    touchy(this.upperWrapper.getCanvas(), removeEvent, 'mousemove', this.onMouseMove);
+    this.dragStatus = DragStatus.Stop;
+
+    this.worker.mouseup();
+    this.emit('mouseup');
+  }
+
+  isOutside(point: Point): boolean {
+    if (
+      point.y < 0 ||
+      point.x < 0 ||
+      point.y > this.lowerWrapper.getHeight() ||
+      point.x > this.lowerWrapper.getWidth()
+    ) {
+      return true;
+    }
+    return false;
+  }
+
+  drawAll(shapes: Array<Shape>, wrapper: CanvasWrapper = this.lowerWrapper) {
+    this.clear(wrapper);
+    for (const shape of shapes) {
+      if (shape.type === 'line') {
+        drawLine(wrapper.getContext(), shape);
+      } else if (shape.type === 'eraser') {
+        drawLine(wrapper.getContext(), shape);
+      } else if (shape.type === 'rect') {
```

Comment:
Can we replace all 'rect's with `ToolType.Rect` ?

Reply:
@hackerwins Thank you for your comment

It is possible to substitute, but it is used in a slightly different meaning, so I am careful.


For `ToolType`
rect, eraser, line, selector, none  ...

For `shape.type`
rect, eraser, line ...